dev-tools/mage/gotest.go

```diff
@@ -0,0 +1,327 @@
+	params := GoTestArgs{
+		TestName:        name,
+		Race:            RaceDetector,
+		Packages:        []string{"./..."},
+		OutputFile:      fileName + ".out",
+		JUnitReportFile: fileName + ".xml",
+	}
+	if TestCoverage {
+		params.CoverageProfileFile = fileName + ".cov"
+	}
+	return params
+}
+
+// DefaultGoTestUnitArgs returns a default set of arguments for running
+// all unit tests. We tag unit test files with '!integration'.
+func DefaultGoTestUnitArgs() GoTestArgs { return makeGoTestArgs("Unit") }
+
+// DefaultGoTestIntegrationArgs returns a default set of arguments for running
+// all integration tests. We tag integration test files with 'integration'.
+func DefaultGoTestIntegrationArgs() GoTestArgs {
+	args := makeGoTestArgs("Integration")
+	args.Tags = append(args.Tags, "integration")
+	return args
+}
+
+// GoTest invokes "go test" and reports the results to stdout. It returns an
+// error if there was any failuring executing the tests or if there were any
+// test failures.
+func GoTest(ctx context.Context, params GoTestArgs) error {
+	fmt.Println(">> go test:", params.TestName, "Testing")
+
+	// Build args list to Go.
+	args := []string{"test", "-v"}
+	if len(params.Tags) > 0 {
+		args = append(args, "-tags", strings.Join(params.Tags, " "))
+	}
+	if params.CoverageProfileFile != "" {
+		params.CoverageProfileFile = createDir(filepath.Clean(params.CoverageProfileFile))
+		args = append(args,
+			"-covermode=atomic",
+			"-coverprofile="+params.CoverageProfileFile,
+		)
+	}
+	args = append(args, params.ExtraFlags...)
+	args = append(args, params.Packages...)
+
+	goTest := makeCommand(ctx, params.Env, "go", args...)
+
+	// Wire up the outputs.
+	bufferOutput := new(bytes.Buffer)
+	outputs := []io.Writer{bufferOutput}
+	if mg.Verbose() {
+		outputs = append(outputs, os.Stdout)
+	}
+	if params.OutputFile != "" {
+		fileOutput, err := os.Create(createDir(params.OutputFile))
+		if err != nil {
+			return errors.Wrap(err, "failed to create go test output file")
+		}
+		defer fileOutput.Close()
+		outputs = append(outputs, fileOutput)
+	}
+	output := io.MultiWriter(outputs...)
+	goTest.Stdout = output
+	goTest.Stderr = output
+
+	// Execute 'go test' and measure duration.
+	start := time.Now()
+	err := goTest.Run()
+	duration := time.Since(start)
+	var goTestErr *exec.ExitError
+	if err != nil {
+		// Command ran.
+		exitErr, ok := err.(*exec.ExitError)
+		if !ok {
+			return errors.Wrap(err, "failed to execute go")
+		}
+
+		// Command ran but failed. Process the output.
+		goTestErr = exitErr
+	}
+
+	// Parse the verbose test output.
+	report, err := parser.Parse(bytes.NewBuffer(bufferOutput.Bytes()), BeatName)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse go test output")
+	}
+	if goTestErr != nil && len(report.Packages) == 0 {
+		// No packages were tested. Probably the code didn't compile.
+		fmt.Println(bytes.NewBuffer(bufferOutput.Bytes()).String())
```

Comment:
I wonder if in here we could also use the log package?

Reply:
In `mage` the stdlib `log` package is tied to the `-v` parameter so it gets used for adding debug information to build.

stdout is the main UI for the build tool. So `fmt.Print` is used for messages that always need to be presented to the user. The goal is to keep verbosity down so that real warnings and problems are easily observed.

I think switching to a leveled logger would be best similar to how Gradle handles it.



And I try to only `fmt.Print` when necessary such that the build only gets verbose when there is a failure.